cli/src/pcluster/templates/budget_builder.py

```diff
@@ -60,35 +86,157 @@ def _add_parameters(self, budget, index):
         )
         return cfn_budget
 
+    def _add_budget_notifications(self, budget):
+        """Create the notifications with subscribers for each budget."""
+        notifications_with_subscribers = []
+        for single_notification in budget.notifications_with_subscribers:
+            subscribers = [
+                budgets.CfnBudget.SubscriberProperty(
+                    address=subscriber.address,
+                    subscription_type=subscriber.subscription_type,
+                )
+                for subscriber in single_notification.subscribers
+            ]
 
-def add_budget_notifications(budget):
-    """Create the notifications with subscribers for each budget."""
-    notifications_with_subscribers = []
-    for single_notification in budget.notifications_with_subscribers:
-        notifications_with_subscribers.append(
-            budgets.CfnBudget.NotificationWithSubscribersProperty(
-                notification=budgets.CfnBudget.NotificationProperty(
-                    comparison_operator=single_notification.notification.comparison_operator,
-                    notification_type=single_notification.notification.notification_type,
-                    threshold_type=single_notification.notification.threshold_type,
-                    threshold=single_notification.notification.threshold,
-                ),
-                subscribers=[
+            if single_notification.trigger_fleet_stop:
+                if not self.sns_arn:
+                    self._create_lambda_and_sns()
+                subscribers.append(
                     budgets.CfnBudget.SubscriberProperty(
-                        address=subscriber.address,
-                        subscription_type=subscriber.subscription_type,
+                        address=self.sns_arn,
+                        subscription_type="SNS",
                     )
-                    for subscriber in single_notification.subscribers
+                )
+            notifications_with_subscribers.append(
+                budgets.CfnBudget.NotificationWithSubscribersProperty(
+                    notification=budgets.CfnBudget.NotificationProperty(
+                        comparison_operator=single_notification.notification.comparison_operator,
+                        notification_type=single_notification.notification.notification_type,
+                        threshold_type=single_notification.notification.threshold_type,
+                        threshold=single_notification.notification.threshold,
+                    ),
+                    subscribers=subscribers,
+                )
+            )
+        return notifications_with_subscribers
+
+    def _create_lambda_and_sns(self):
+        """Create an sns topic and a lambda function to stop compute fleet when triggered."""
+        self.sns_topic = sns.Topic(
+            self.scope,
+            "BudgetTopic",
+            display_name="BudgetTriggeredNotification",
```

Comment:
as we discussed, let's test if this name need to be unique for different clusters

Reply:
Naming updated and tested!